network/src/internal/message_handler.rs

```diff
@@ -295,12 +304,26 @@ impl Server {
     /// A node has requested our list of peer addresses.
     /// Send an Address message with our current peer list.
     async fn receive_get_peers(&mut self, _message: GetPeers, channel: Arc<Channel>) -> Result<(), ServerError> {
-        let mut connected_peers = self.context.peer_book.read().await.get_connected();
-
-        // Remove the requester from list of peers
-        connected_peers.remove(&channel.address);
+        // If we received a message, but aren't connected to the peer,
+        // inform the peer book that we found a peer.
+        // The peer book will determine if we have seen the peer before,
+        // and include the peer if it is new.
+        let mut peer_book = self.context.peer_book.write().await;
+        if !peer_book.is_connected(&channel.address) {
+            peer_book.found_peer(&channel.address);
+        }
 
-        channel.write(&Peers::new(connected_peers)).await?;
+        // Broadcast the sanitized list of connected peers back to requesting peer.
+        let mut peers = HashMap::new();
+        for (remote_address, peer_info) in peer_book.get_all_connected().iter() {
```

Comment:
if it is expected that there can be many simultaneously connected peers, the call to `get_all_connected` could be done earlier, the returned value cloned, and the `peer_book` lock manually `drop()`ped right afterwards in order not to keep it open for too long